File: LeetCode/881-boats-to-save-people.c
#include "_leetcode_common.c"

int sorter(const void *a, const void *b) {
  return *(int *) a - *(int *) b;
}

int numRescueBoats(int *people, int peopleSize, int limit) {
  qsort(people, peopleSize, sizeof(*people), sorter);

  int res = 0;

  int l = 0, r = peopleSize - 1;
  while (l <= r) {
    // the lightest person can share with the heaviest person
    if (people[l] + people[r] <= limit)
      ++l;
    // the heaviest person takes a boat
    --r;
    ++res;
  }

  return res;
}

int main(int argc, char *argv[]) {
  return 0;
}

File: LeetCode/881-boats-to-save-people.py
from typing import List


class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        people.sort()

        res = 0
        l, r = 0, len(people) - 1

        while l <= r:
            # the lightest person can share with the heaviest person?
            if people[r] + people[l] <= limit:
                l += 1
            # handle the heaviest person
            r -= 1
            res += 1

        return res


s = Solution()

res = s.numRescueBoats([3, 2, 2, 1], 3)
print(f"{res = }")
